File: frontend/apps/ios/Pantopus/Features/Homes/Guests/AddGuestFormViewModel.swift
//
//  AddGuestFormViewModel.swift
//  Pantopus
//
//  A13.1 — Backs the Add Guest form (issue a short-term guest pass for a
//  home). Built on the shared `FormShell` archetype: name + contact +
//  welcome are tracked as `FormFieldState`s; duration (single-select) and
//  allowed-areas (multi-select) are enum-ish chip selections held
//  directly.
//
//  No backend: issuing a pass is simulated locally — `submit()` flips a
//  brief saving state, raises a success toast ("Pass sent to <name>"),
//  and signals `shouldDismiss` so the host pops the modal.
//

import Foundation
import Observation
import SwiftUI

@Observable
@MainActor
public final class AddGuestFormViewModel {
    // MARK: - Bound state

    public private(set) var nameField: FormFieldState
    public private(set) var contactField: FormFieldState
    public private(set) var welcomeField: FormFieldState

    /// Selected duration chip id (single-select). `nil` until the user
    /// picks one. `"custom"` opens the date-range sheet.
    public var duration: String?

    /// Selected allowed-area chip ids (multi-select, optional).
    public var selectedAreas: Set<String> = []

    /// Custom date range, populated when the user commits the picker.
    public private(set) var customStart: Date?
    public private(set) var customEnd: Date?

    public private(set) var isSaving = false
    public var toast: ToastMessage?
    public private(set) var shouldDismiss = false

    // MARK: - Inputs

    public let homeId: String
    public let homeContext: AddGuestSampleData.HomeContext
    public let durationOptions = AddGuestSampleData.durationOptions
    public let areaOptions = AddGuestSampleData.areaOptions
    public let welcomeMaxLength = AddGuestSampleData.welcomeMaxLength

    private let onSent: (String) -> Void

    public init(homeId: String, onSent: @escaping (String) -> Void = { _ in }) {
        self.homeId = homeId
        homeContext = AddGuestSampleData.homeContext(for: homeId)
        nameField = FormFieldState(id: "name", originalValue: "")
        contactField = FormFieldState(id: "contact", originalValue: "")
        welcomeField = FormFieldState(id: "welcome", originalValue: "")
        self.onSent = onSent
    }

    // MARK: - Aggregate

    /// Required: name non-empty, contact valid (email OR phone), duration
    /// chosen. Areas + welcome are optional.
    public var isValid: Bool {
        !trimmedName.isEmpty
            && Self.isContactValid(contactField.value)
            && duration != nil
    }

    /// Any input touched — drives the dirty-close confirm in `FormShell`.
    public var isDirty: Bool {
        !trimmedName.isEmpty
            || !contactField.value.isEmpty
            || duration != nil
            || !selectedAreas.isEmpty
            || !welcomeField.value.isEmpty
    }

    // MARK: - Field states (drive the text-field visuals)

    public var nameFieldState: PantopusFieldState {
        (nameField.touched && !trimmedName.isEmpty) ? .valid : .default
    }

    public var contactFieldState: PantopusFieldState {
        if let error = contactField.error { return .error(error) }
        if contactField.touched, Self.isContactValid(contactField.value) { return .valid }
        return .default
    }

    // MARK: - Hints

    /// Italic helper under the duration chips. Mirrors the design copy.
    public var durationHint: String {
        switch duration {
        case "2h": "Two-hour pass · auto-revokes after"
        case "today": "Today until 11:59 PM · auto-revokes after"
        case "weekend": "Sat 12:00 AM → Sun 11:59 PM · auto-revokes after"
        case AddGuestSampleData.durationCustomId:
            if let start = customStart, let end = customEnd {
                "\(Self.format(start)) → \(Self.format(end)) · auto-revokes after"
            } else {
                "Pick a custom date range"
            }
        default: "Pick how long the pass is good for."
        }
    }

    /// Italic helper under the allowed-areas chips.
    public var areasHint: String {
        guard !selectedAreas.isEmpty else { return "Front door only, unless you add more." }
        let possessive = firstName.map { "\($0)'s" } ?? "Their"
        return "\(possessive) pass unlocks only what you pick."
    }

    /// First word of the entered name, if any — used for the toast and
    /// the areas-hint possessive.
    public var firstName: String? {
        trimmedName.split(separator: " ").first.map(String.init)
    }

    private var trimmedName: String {
        nameField.value.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    // MARK: - Field updates

    public func updateName(_ value: String) {
        nameField.value = value
        nameField.touched = true
    }

    public func updateContact(_ value: String) {
        contactField.value = value
        contactField.touched = true
        contactField.error = Self.validateContact(value)
    }

    public func updateWelcome(_ value: String) {
        welcomeField.value = String(value.prefix(welcomeMaxLength))
        welcomeField.touched = true
    }

    // MARK: - Custom duration range

    public func setCustomRange(_ start: Date, _ end: Date) {
        customStart = start
        customEnd = end
        duration = AddGuestSampleData.durationCustomId
    }

    public func clearCustomRange() {
        customStart = nil
        customEnd = nil
        if duration == AddGuestSampleData.durationCustomId { duration = nil }
    }

    // MARK: - Submit

    public func submit() async {
        guard isValid, !isSaving else { return }
        isSaving = true
        // No backend in the repo — simulate the issue-pass round trip so
        // the CTA spinner and dismiss flow behave like the real thing.
        try? await Task.sleep(nanoseconds: 350_000_000)
        isSaving = false
        let name = firstName ?? "your guest"
        toast = ToastMessage(text: "Pass sent to \(name)", kind: .success)
        onSent(name)
        shouldDismiss = true
    }

    public func acknowledgeDismiss() {
        shouldDismiss = false
    }

    // MARK: - Validation (static — testable without the view layer)

    /// Returns an error string when a non-empty contact is neither a
    /// valid email nor a valid phone; `nil` when empty (required-ness is
    /// enforced by `isValid`) or valid.
    static func validateContact(_ raw: String) -> String? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return nil }
        return isContactValid(trimmed) ? nil : "Enter a valid email or phone number."
    }

    static func isContactValid(_ raw: String) -> Bool {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return false }
        return matchesEmail(trimmed) || matchesPhone(trimmed)
    }

    static func matchesEmail(_ value: String) -> Bool {
        let pattern = "^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$"
        return value.range(of: pattern, options: .regularExpression) != nil
    }

    static func matchesPhone(_ value: String) -> Bool {
        let allowed = CharacterSet(charactersIn: "0123456789 +()-.")
        guard value.unicodeScalars.allSatisfy({ allowed.contains($0) }) else { return false }
        let digits = value.filter(\.isNumber)
        return digits.count >= 7 && digits.count <= 15
    }

    private static func format(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "MMM d"
        return formatter.string(from: date)
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/homes/guests/AddGuestFormViewModel.kt
@file:Suppress("PackageNaming", "MagicNumber", "TooManyFunctions")

package app.pantopus.android.ui.screens.homes.guests

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.ui.components.ChipPickerOption
import app.pantopus.android.ui.screens.shared.form.FormFieldState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/** Nav-arg key for the home id consumed via [SavedStateHandle]. */
const val ADD_GUEST_HOME_ID_KEY = "homeId"

/** Tone + text payload the form turns into a transient toast. */
data class GuestToast(
    val text: String,
    val isError: Boolean,
)

/**
 * Aggregate UI state for the Add Guest form. Mirrors iOS
 * `AddGuestFormViewModel`'s exposed state: name / contact / welcome are
 * tracked as [FormFieldState]s, duration is a single-select chip id, and
 * allowed-areas is a set of chip ids.
 */
data class AddGuestUiState(
    val homeTitle: String = "",
    val homeSubtitle: String = "",
    val nameField: FormFieldState = FormFieldState(id = "name"),
    val contactField: FormFieldState = FormFieldState(id = "contact"),
    val welcomeField: FormFieldState = FormFieldState(id = "welcome"),
    val duration: String? = null,
    val selectedAreas: Set<String> = emptySet(),
    val customStartLabel: String? = null,
    val customEndLabel: String? = null,
    val isSaving: Boolean = false,
    val toast: GuestToast? = null,
    val shouldDismiss: Boolean = false,
) {
    val durationOptions: List<ChipPickerOption> get() = AddGuestSampleData.durationOptions
    val areaOptions: List<ChipPickerOption> get() = AddGuestSampleData.areaOptions
    val welcomeMaxLength: Int get() = AddGuestSampleData.WELCOME_MAX_LENGTH

    /** First word of the entered name, if any. */
    val firstName: String?
        get() = nameField.value.trim().split(" ").firstOrNull()?.takeIf { it.isNotEmpty() }

    /** Required: name non-empty, contact valid (email OR phone), duration chosen. */
    val isValid: Boolean
        get() =
            nameField.value.trim().isNotEmpty() &&
                isGuestContactValid(contactField.value) &&
                duration != null

    /** Any input touched — drives the dirty-close confirm in `FormShell`. */
    val isDirty: Boolean
        get() =
            nameField.value.trim().isNotEmpty() ||
                contactField.value.isNotEmpty() ||
                duration != null ||
                selectedAreas.isNotEmpty() ||
                welcomeField.value.isNotEmpty()

    /** Italic helper under the duration chips. Mirrors the design copy. */
    val durationHint: String
        get() =
            when (duration) {
                "2h" -> "Two-hour pass · auto-revokes after"
                "today" -> "Today until 11:59 PM · auto-revokes after"
                "weekend" -> "Sat 12:00 AM → Sun 11:59 PM · auto-revokes after"
                AddGuestSampleData.DURATION_CUSTOM_ID ->
                    if (customStartLabel != null && customEndLabel != null) {
                        "$customStartLabel → $customEndLabel · auto-revokes after"
                    } else {
                        "Pick a custom date range"
                    }
                else -> "Pick how long the pass is good for."
            }

    /** Italic helper under the allowed-areas chips. */
    val areasHint: String
        get() =
            if (selectedAreas.isEmpty()) {
                "Front door only, unless you add more."
            } else {
                val possessive = firstName?.let { "$it's" } ?: "Their"
                "$possessive pass unlocks only what you pick."
            }
}

/**
 * A13.1 — Add Guest form view-model. No backend in the repo, so issuing a
 * pass is simulated locally: [submit] flips a brief saving state, raises a
 * success toast ("Pass sent to <name>"), and signals [AddGuestUiState.shouldDismiss]
 * so the host pops the modal.
 */
@HiltViewModel
class AddGuestFormViewModel
    @Inject
    constructor(
        savedStateHandle: SavedStateHandle,
    ) : ViewModel() {
        private val homeId: String = savedStateHandle.get<String>(ADD_GUEST_HOME_ID_KEY) ?: ""

        private val _state =
            MutableStateFlow(
                AddGuestSampleData.homeContext(homeId).let { ctx ->
                    AddGuestUiState(homeTitle = ctx.title, homeSubtitle = ctx.subtitle)
                },
            )
        val state: StateFlow<AddGuestUiState> = _state.asStateFlow()

        fun updateName(value: String) {
            _state.update { it.copy(nameField = it.nameField.copy(value = value, touched = true)) }
        }

        fun updateContact(value: String) {
            _state.update {
                it.copy(
                    contactField =
                        it.contactField.copy(
                            value = value,
                            touched = true,
                            error = validateGuestContact(value),
                        ),
                )
            }
        }

        fun updateWelcome(value: String) {
            val clipped = value.take(AddGuestSampleData.WELCOME_MAX_LENGTH)
            _state.update { it.copy(welcomeField = it.welcomeField.copy(value = clipped, touched = true)) }
        }

        fun setDuration(id: String?) {
            _state.update { it.copy(duration = id) }
        }

        fun setAreas(areas: Set<String>) {
            _state.update { it.copy(selectedAreas = areas) }
        }

        fun setCustomRange(
            startLabel: String,
            endLabel: String,
        ) {
            _state.update {
                it.copy(
                    duration = AddGuestSampleData.DURATION_CUSTOM_ID,
                    customStartLabel = startLabel,
                    customEndLabel = endLabel,
                )
            }
        }

        fun clearCustomRange() {
            _state.update {
                val nextDuration =
                    if (it.duration == AddGuestSampleData.DURATION_CUSTOM_ID) null else it.duration
                it.copy(duration = nextDuration, customStartLabel = null, customEndLabel = null)
            }
        }

        fun submit() {
            val current = _state.value
            if (!current.isValid || current.isSaving) return
            _state.update { it.copy(isSaving = true) }
            viewModelScope.launch {
                // No backend in the repo — simulate the issue-pass round trip.
                delay(350)
                val name = _state.value.firstName ?: "your guest"
                _state.update {
                    it.copy(
                        isSaving = false,
                        toast = GuestToast("Pass sent to $name", isError = false),
                        shouldDismiss = true,
                    )
                }
            }
        }

        fun dismissToast() {
            _state.update { it.copy(toast = null) }
        }

        fun acknowledgeDismiss() {
            _state.update { it.copy(shouldDismiss = false) }
        }
    }

// ─── Validation — top-level so the UiState getters + VM share them ──────

internal fun validateGuestContact(raw: String): String? {
    val trimmed = raw.trim()
    if (trimmed.isEmpty()) return null
    return if (isGuestContactValid(trimmed)) null else "Enter a valid email or phone number."
}

internal fun isGuestContactValid(raw: String): Boolean {
    val trimmed = raw.trim()
    if (trimmed.isEmpty()) return false
    return matchesGuestEmail(trimmed) || matchesGuestPhone(trimmed)
}

private val guestEmailRegex = Regex("^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$")
private val guestPhoneAllowed = Regex("^[0-9 +()\\-.]+$")

private fun matchesGuestEmail(value: String): Boolean = guestEmailRegex.matches(value)

private fun matchesGuestPhone(value: String): Boolean {
    if (!guestPhoneAllowed.matches(value)) return false
    val digits = value.count { it.isDigit() }
    return digits in 7..15
}
